crates/viewer/re_space_view_dataframe/src/dataframe_ui.rs

```diff
@@ -0,0 +1,350 @@
+use std::collections::BTreeMap;
+use std::ops::Range;
+
+use anyhow::Context;
+use itertools::Itertools;
+
+use re_chunk_store::{ColumnDescriptor, LatestAtQuery, RowId};
+use re_dataframe::{LatestAtQueryHandle, RangeQueryHandle, RecordBatch};
+use re_log_types::{EntityPath, TimeInt, Timeline};
+use re_types_core::ComponentName;
+use re_ui::UiExt as _;
+use re_viewer_context::ViewerContext;
+
+use crate::display_record_batch::{DisplayRecordBatch, DisplayRecordBatchError};
+
+/// Display a dataframe table for the provided query.
+pub(crate) fn dataframe_ui<'a>(
+    ctx: &ViewerContext<'_>,
+    ui: &mut egui::Ui,
+    query: impl Into<QueryHandle<'a>>,
+) {
+    dataframe_ui_impl(ctx, ui, &query.into());
+}
+
+/// A query handle for either a latest-at or range query.
+pub(crate) enum QueryHandle<'a> {
+    LatestAt(LatestAtQueryHandle<'a>),
+    Range(RangeQueryHandle<'a>),
+}
+
+impl QueryHandle<'_> {
+    fn schema(&self) -> &[ColumnDescriptor] {
+        match self {
+            QueryHandle::LatestAt(query_handle) => query_handle.schema(),
+            QueryHandle::Range(query_handle) => query_handle.schema(),
+        }
+    }
+
+    fn num_rows(&self) -> u64 {
+        match self {
+            QueryHandle::LatestAt(_) => 1,
+            QueryHandle::Range(query_handle) => query_handle.num_rows(),
+        }
+    }
+
+    fn get(&self, start: u64, num_rows: u64) -> Vec<RecordBatch> {
+        match self {
+            QueryHandle::LatestAt(query_handle) => {
+                vec![query_handle.get()]
+            }
+            QueryHandle::Range(query_handle) => query_handle.get(start, num_rows),
+        }
+    }
+
+    fn timeline(&self) -> Timeline {
+        match self {
+            QueryHandle::LatestAt(query_handle) => query_handle.query().timeline,
+            QueryHandle::Range(query_handle) => query_handle.query().timeline,
+        }
+    }
+}
+
+impl<'a> From<LatestAtQueryHandle<'a>> for QueryHandle<'a> {
+    fn from(query_handle: LatestAtQueryHandle<'a>) -> Self {
+        QueryHandle::LatestAt(query_handle)
+    }
+}
+
+impl<'a> From<RangeQueryHandle<'a>> for QueryHandle<'a> {
+    fn from(query_handle: RangeQueryHandle<'a>) -> Self {
+        QueryHandle::Range(query_handle)
+    }
+}
+
+/// This structure maintains the data for displaying rows in a table.
+///
+/// Row data is stored in a bunch of [`DisplayRecordBatch`], which are created from
+/// [`RecordBatch`]s. We also maintain a mapping for each row number to the corresponding record
+/// batch and the index inside it.
+struct RowsDisplayData {
+    /// The [`DisplayRecordBatch`]s to display.
+    display_record_batches: Vec<DisplayRecordBatch>,
+
+    /// For each row to be displayed, the batch index and the row index with that batch.
+    batch_and_row_from_row_nr: BTreeMap<u64, (usize, usize)>,
```

Comment:
Let's make this self-documenting instead (yes, I know I was the one who wrote this code - I can fix it)

```rs
struct BatchRef {
    batch_index: usize,
    row_index: u64,
}
```

```suggestion
    batch_and_row_from_row_nr: BTreeMap<u64, BatchRef>,
```